common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -59,6 +59,27 @@ public CalendarInterval subtract(CalendarInterval that) {
     return new CalendarInterval(months, days, microseconds);
   }
 
+  private static CalendarInterval fromDoubles(
+      double monthsWithFraction,
+      double daysWithFraction,
+      double microsWithFraction) {
+    int truncatedMonths = Math.toIntExact((long)(monthsWithFraction));
+    // Using 30 days per month as PostgreSQL does.
+    double days = daysWithFraction + 30 * (monthsWithFraction - truncatedMonths);
+    int truncatedDays = Math.toIntExact((long)days);
+    double micros = microsWithFraction + MICROS_PER_DAY * (days - truncatedDays);
+    long truncatedMicros = Math.round(micros);
+    return new CalendarInterval(truncatedMonths, truncatedDays, truncatedMicros);
+  }
+
+  public CalendarInterval multiply(double num) {
```

Comment:
shall we use decimal instead? double is approximate value and we may truncate.

Reply:
- In that case, our implementation will deviate from postgesql which uses double internally. At the moment, we return the same results as postgresql (or I haven't found yet the case when the results are different).
- most likely, it will be slower

If it is ok, I will switch to decimals.

Reply:
OK let's use double if it's what pgsql uses.

Can we move the `add`, `subtract`, `multiply` and `divide` to `IntervalUtils`? In case we want to change them in the future.

Reply:
> Can we move the add, subtract ... to IntervalUtils?

Do you want to move `+` and `-` in this PR? I just want to double check this because those methods are not related to this PR directly.

Reply:
both are fine. We can move them tother, or have a followup PR to move +/-

Reply:
Double is not big enough to support the average aggrege for interval https://github.com/apache/spark/pull/26347, I prefer decimal personally

Reply:
> Double is not big enough to support the average aggregate ...

@yaooqinn Could you explain what do you mean? I could image that `double` is not precise enough but not big though ...

Reply:
yea, precise, not big.